ast/clone.go

```diff
@@ -0,0 +1,302 @@
+/*
+Copyright 2018 Google Inc. All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package ast
+
+import (
+	"fmt"
+	"reflect"
+)
+
+// Updates fields of specPtr to point to deep clones.
+func cloneForSpec(specPtr *ForSpec) {
+	clone(&specPtr.Expr)
+	oldOuter := specPtr.Outer
+	specPtr.Outer = new(ForSpec)
+	*specPtr.Outer = *oldOuter
+	cloneForSpec(specPtr.Outer)
+	for i := range specPtr.Conditions {
+		clone(&specPtr.Conditions[i].Expr)
+	}
+}
+
+// Updates fields of params to point to deep clones.
+func cloneParameters(params *Parameters) {
+	if params == nil {
+		return
+	}
+	params.Optional = append(make([]NamedParameter, 0), params.Optional...)
+	for i := range params.Optional {
+		clone(&params.Optional[i].DefaultArg)
+	}
+}
+
+// Updates fields of field to point to deep clones.
+func cloneField(field *ObjectField) {
+	if field.Method != nil {
+		field.Method = Clone(field.Method).(*Function)
+	}
+
+	oldParams := field.Params
+	if oldParams != nil {
+		field.Params = new(Parameters)
+		*field.Params = *oldParams
+	}
+	cloneParameters(field.Params)
+
+	clone(&field.Expr1)
+	clone(&field.Expr2)
+	clone(&field.Expr3)
+}
+
+// Updates fields of field to point to deep clones.
+func cloneDesugaredField(field *DesugaredObjectField) {
+	clone(&field.Name)
+	clone(&field.Body)
+}
+
+// Updates the NodeBase fields of astPtr to point to deep clones.
+func cloneNodeBase(astPtr Node) {
+	if astPtr.Context() != nil {
+		newContext := new(string)
+		*newContext = *astPtr.Context()
+		astPtr.SetContext(newContext)
+	}
+	astPtr.SetFreeVariables(append(make(Identifiers, 0), astPtr.FreeVariables()...))
+}
+
+// Updates *astPtr to point to a deep clone of what it originally pointed at.
```

Comment:
"In-place clone" is a bit weird. I see it saves quite a lot of boilerplate, so it's ok. I'm worried it could be a bit bugprone - it's easy to forget to clone a field. Checking that the result is a proper tree would help.

The check could be something along these lines:
```
func check(root *Node) {
    visited := make(map [Node*]bool);
    func dfs(node *Node) {
          visited[node] = true;
          for child := range children(node) {
              if (visited[child]) {
                    error("Not a tree");
                    return;
              }
              dfs(node);
          }
          dfs(root);
    }
```

Reply:
Yeah, although even the most explicit option: `&Foo { x: other.x, y: clone(other.y), ... }` is liable to have you forget to clone stuff because you didn't realise it's a pointer (easy given the typedefs, interfaces). I think it's OK because we'll want that checker anyway.  It's tricky to give a good error message pointing to where in the tree the duplication occurred, though!

Reply:
I guess the path of types (concrete types, not Node) is easy enough to provide (just need an explicit stack of Node and use reflection) and specific enough to find the problem.
visited needs to map Node to the path though, so we can say where it was found elsewhere.

Reply:
Perhaps it's possible to record the path or reuse the context for that.

It's only an assertion for tests, not a user-facing message, so I wouldn't spend too much time on making it better.